extensions/vi-mode/word.lisp

```diff
@@ -1,57 +1,75 @@
 (defpackage :lem-vi-mode/word
-  (:use :cl :lem)
+  (:use :cl
+        :lem)
+  (:import-from :lem-vi-mode/options
+                :vi-option-raw-value)
   (:export :forward-word-begin
+           :forward-word-end
            :backward-word-begin
-           :forward-word-end))
+           :word-char-p
+           :char-type
+           :broad-char-type
+           :blank-char-p))
 (in-package :lem-vi-mode/word)
 
-(defun hiragana-p (c)
-  (<= #x3042 (char-code c) #x3093))
-
-(defun katakana-p (c)
-  (<= #x30A2 (char-code c) #x30F3))
-
-(defun kanji-p (c)
-  (or (<= #x4E00 (char-code c) #x9FFF)
-      (find c "仝々〆〇ヶ")))
-
-(defun paren-p (c)
-  (or (syntax-open-paren-char-p c)
-      (syntax-closed-paren-char-p c)))
-
-(defun word-type (c largep)
-  (flet ((w (small large) (if largep large small)))
-    (cond ((syntax-space-char-p c)
-           (w :space :space))
-          ((hiragana-p c)
-           (w :hiragana :symbol))
-          ((katakana-p c)
-           (w :katakana :symbol))
-          ((kanji-p c)
-           (w :kanji :symbol))
-          ((or (syntax-open-paren-char-p c)
-               (syntax-closed-paren-char-p c))
-           (w :paren :symbol))
-          (t
-           (w :symbol :symbol)))))
-
-(defun forward-word-begin (point n largep)
-  (loop :repeat n
-        :until (end-buffer-p point)
-        :do (let ((wt (word-type (character-at point) largep)))
-              (skip-chars-forward point (lambda (c) (eql wt (word-type c largep))))
-              (skip-whitespace-forward point))))
-
-(defun backward-word-begin (point n largep)
-  (loop :repeat n
-        :until (start-buffer-p point)
-        :do (skip-whitespace-backward point)
-            (let ((wt (word-type (character-at point -1) largep)))
-              (skip-chars-backward point (lambda (c) (eql wt (word-type c largep)))))))
-
-(defun forward-word-end (point n largep)
-  (loop :repeat n
-        :until (end-buffer-p point)
-        :do (skip-whitespace-forward point)
-            (let ((wt (word-type (character-at point) largep)))
-              (skip-chars-forward point (lambda (c) (eql wt (word-type c largep)))))))
+(defun word-char-p (char)
+  (funcall (cdr (vi-option-raw-value "iskeyword"))
+           char))
+
+(defun blank-char-p (char)
+  (and (member char '(#\Newline #\Space #\Tab))
+       t))
+
+(defun non-blank-char-p (char)
+  (not (blank-char-p char)))
+
+(defun char-type (char)
+  (when char
+    (cond
+      ((word-char-p char) :word)
+      ((blank-char-p char) :blank)
+      (t :non-word))))
+
+(defun broad-char-type (char)
+  (when char
+    (cond
+      ((blank-char-p char) :blank)
+      (t :word))))
+
+(defun forward-word-begin (char-type-fn &optional (point (current-point)))
+  (symbol-macrolet ((point-char (character-at point)))
```

Comment:
This is a very cool way to write, but I am concerned that others will see this and copy it elsewhere.
What about using labels/flet?

Reply:
Okay, I replaced them with simple `flet`s.
Actually, they didn't need symbol macros, I don't know I wrote them with it.